File: src/logic/indicators.js
/**
 * Builds sorted, filtered indicator data for the Driving View.
 */

const Indicators = (() => {
  // Viewport-tiered NAV display cap — small phone screens stay glanceable
  // with fewer indicators; larger tablet/infotainment-style displays have
  // room for more before it becomes clutter.
  const NAV_CAP_SMALL_MAX_WIDTH  = 500; // px, exclusive
  const NAV_CAP_MEDIUM_MAX_WIDTH = 900; // px, inclusive
  const NAV_CAP_SMALL  = 5;
  const NAV_CAP_MEDIUM = 7;
  const NAV_CAP_LARGE  = 10;

  /**
   * @param {number} width  Viewport width in px (real or emulated).
   * @returns {number} Max NAV indicators to show at that width.
   */
  function capForViewportWidth(width) {
    if (width < NAV_CAP_SMALL_MAX_WIDTH) return NAV_CAP_SMALL;
    if (width <= NAV_CAP_MEDIUM_MAX_WIDTH) return NAV_CAP_MEDIUM;
    return NAV_CAP_LARGE;
  }

  /**
   * True polar plot range — the distance (nm) that maps to the full
   * available radius from the anchor. Reuses Relevance's own dead-ahead
   * teardrop range rather than a separate made-up number, so the plotted
   * scale actually means something: an aircraft at the plot's outer edge is
   * an aircraft at (or beyond) the edge of relevance for that bearing.
   *
   * This is `rangeExtensionCapNm` (50), not the base `rMaxNm` (15) —
   * updated 2026-08-21 alongside the contrail-visibility work. Before that,
   * relevance never reached past 15nm for anything, so tying the plot's
   * scale to 15 and to relevance's own max were the same thing. Now a
   * high-altitude aircraft can be relevant out to 50nm, and leaving the
   * plot's own scale capped at 15 meant every one of those aircraft —
   * regardless of whether it was 16nm or 46nm away — clamped to the exact
   * same outer-edge radius: no differentiation, no correlation with the
   * range rings, and (with several of them at once) no room to avoid their
   * labels overlapping either. Reported directly against the deployed app.
   */
  const POLAR_MAX_RANGE_NM = Relevance.DEFAULTS.rangeExtensionCapNm;

  /**
   * Ring band boundaries (nm) for the plot's non-linear distance scale —
   * see Geo.bandedRadiusFraction(). Each band gets an equal slice of the
   * available radius regardless of its real nm width, so close traffic
   * (where resolving "is this one a bit closer than that one" actually
   * matters) gets more usable screen space than a strictly linear scale
   * would give it, while distant traffic still visibly separates into "how
   * far, roughly" bands instead of a single distant blob. The final band
   * (15 to POLAR_MAX_RANGE_NM/50) is deliberately wide — it only ever holds
   * high-altitude/contrail-relevant traffic (see relevance.js), which is
   * inherently sparser than low-altitude local traffic, so one wide band
   * is enough to keep them from collapsing onto the exact same point
   * without needing the same fine-grained resolution the close bands have.
   * The map's own range rings (EosMap.updateRangeRings, real geo circles)
   * are drawn at these same nm values, but a 50nm ring's true geo radius
   * doesn't fit on screen at RAW's zoom regardless — same as the 15nm ring
   * already sometimes doesn't (see "Range rings" below) — so it's harmless
   * that it's declared here even though it'll rarely if ever actually be
   * visible; the banded *plot position* for far traffic is the part that
   * matters.
   */
  const RING_BANDS_NM = [2, 5, 10, 15, POLAR_MAX_RANGE_NM];

  /**
   * Half-angle (deg) of RAW mode's forward field of view — 75° either side
   * of dead-ahead, ~150° total, matching a real TCAS/ND reference photo
   * (which shows a forward arc, never a full 360° sweep) and RAW's own
   * rationale: it exists to show what's ahead while driving, not a
   * behind-the-vehicle picture. Passed through as userState.fovHalfAngleDeg
   * by RAW's caller only — Hybrid's edge indicators have no "round display"
   * to fit an arc into and keep the full teardrop Relevance computes.
   */
  const FOV_HALF_ANGLE_DEG = 75;

  /**
   * Shared per-aircraft computation used by both build() and buildAll() —
   * bearing/distance/visibility/relevance/polar screen position/direction-
   * of-travel, for every aircraft that passes the hard staleness cutoff.
   * Does NOT filter by relevance or suppression; callers decide what to do
   * with that.
   */
  function _computeAll(aircraftList, userState, staleThresholdSeconds) {
    const { lat, lon, heading, viewportWidth, viewportHeight, metar } = userState;
    // Must match what the camera actually used for this frame (see
    // CameraController/geo.js's projectToPolarPosition doc comment) or the
    // plotted origin silently drifts from where the user marker and range
    // rings really are. anchorY defaults to projectToPolarPosition's own
    // default (0.8) if the caller didn't evaluate a camera frame yet (e.g.
    // before the first followNav() call); safeInset likewise falls back to
    // its own default.
    const anchorY = userState.anchorY;
    const fovHalfAngleDeg = userState.fovHalfAngleDeg;
    // RAW's plot lives inside a 1:1 square sub-region of the real viewport
    // (Geo.computeSquarePlotLayout), not the full viewport itself — these
    // default back to the plain viewport with no offset when unset, so
    // Hybrid (which never sets them) is completely unaffected.
    const plotWidth = userState.plotWidth != null ? userState.plotWidth : viewportWidth;
    const plotHeight = userState.plotHeight != null ? userState.plotHeight : viewportHeight;
    const plotOffsetX = userState.plotOffsetX || 0;
    const plotOffsetY = userState.plotOffsetY || 0;
    // safeInset means "real chrome height to stay clear of" for Hybrid's
    // unrestricted, full-viewport teardrop — but the square ALREADY
    // excludes all real chrome from its own bounds (see app.js's
    // _rawChromeInsets), so reusing that same ~60-100px value as margin
    // WITHIN the square would carve out a needlessly huge chunk of a
    // region that's already chrome-free. plotSafeInset (set only for RAW)
    // is a small fixed in-square margin instead; falling back to safeInset
    // keeps Hybrid's existing behaviour exactly as it was.
    const safeInset = userState.plotSafeInset != null ? userState.plotSafeInset : userState.safeInset;
    // The user-selectable ND-style range knob (app.js's selectedRangeIndex)
    // overrides which band boundaries are "in play" for THIS render — e.g.
    // dialled down to 10nm, bandsNm becomes [2,5,10] instead of the full
    // [2,5,10,15,50], so bandedRadiusFraction's own existing clamp-to-1.0
    // behaviour (anything at/beyond the last band plots at the outer edge)
    // is what naturally produces the "suppressed edge dot" position for
    // traffic beyond the selected range — no separate geometry needed for
    // that, see app.js's refreshIndicators. Falls back to the full array
    // for Hybrid/LogPanel callers, which never set this.
    const bandsNm = userState.plotBandsNm || RING_BANDS_NM;

    return aircraftList
      .filter(a => a.lastSeenSeconds < staleThresholdSeconds * 3) // hard cut
      .map(a => {
        const bearing = Geo.calculateBearing(lat, lon, a.lat, a.lon);
        const distanceNm = Geo.calculateDistanceNm(lat, lon, a.lat, a.lon);
        const vis = Visibility.estimate(lat, lon, a, metar);
        const relativeBearing = Geo.calculateRelativeBearing(bearing, heading);
        const relevance = Relevance.evaluate(userState, a, relativeBearing, vis);
        // Slant range (not flat horizontal distance) — the same figure
        // Relevance itself compares against the teardrop, so an aircraft's
        // plotted radius agrees with whether it's near the edge of relevance.
        // null when fovHalfAngleDeg is set and this bearing falls outside
        // RAW's forward field of view — callers must skip rendering those
        // (still relevant/tracked for Hybrid/logging, just not drawable
        // inside RAW's arc-restricted plot).
        const pos = Geo.projectToPolarPosition(relativeBearing, vis.slantRangeNm, plotWidth, plotHeight, bandsNm, anchorY, safeInset, fovHalfAngleDeg, plotOffsetX, plotOffsetY);
        const x = pos ? pos.x : null;
        const y = pos ? pos.y : null;
        // Direction-of-travel indicator — the aircraft's own ground track,
        // expressed relative to the observer's heading-up view the same way
        // relativeBearing expresses the aircraft's *position*. null when the
        // aircraft isn't transmitting a track (no arrow drawn for those).
        const relativeTrackDeg = a.trackDeg != null ? Geo.calculateRelativeBearing(a.trackDeg, heading) : null;
        const isStale = a.lastSeenSeconds > staleThresholdSeconds;

        return {
          aircraft: a,
          bearing,
          distanceNm,
          relativeBearing,
          relativeTrackDeg,
          vis,
          relevance,
          x, y,
          isStale,
        };
      });
  }

  /**
   * Given the full aircraft list and user state, return every relevant
   * aircraft, sorted best-candidate-first. Aircraft that Relevance.evaluate()
   * rules out (roughly: behind the user, not close, not converging into
   * view) never reach the sort stage at all — this is a TCAS-style
   * relevance gate, not just a visibility ranking.
   *
   * Deliberately unpaginated — the caller decides how many to actually
   * display (via capForViewportWidth) and which page, since that's
   * display/interaction state, not something this pure function should own.
   *
   * userState: { lat, lon, heading, speedMph, viewportWidth, viewportHeight, metar }
   *   metar (optional): current MetarProvider.getCached() snapshot, passed
   *   straight through to Visibility.estimate() — see there for what it does.
   * @param {Set<string>} [suppressedHexes]  Aircraft hex codes to exclude regardless of
   *   relevance (manually dismissed via the popup's Suppress button). Applies uniformly —
   *   there's no relevance reason exempt from suppression, including overhead/close cases.
   */
  function build(aircraftList, userState, staleThresholdSeconds, suppressedHexes) {
    const withMeta = _computeAll(aircraftList, userState, staleThresholdSeconds)
      .filter(item => !suppressedHexes || !suppressedHexes.has(item.aircraft.hex))
      .filter(item => item.relevance.relevant);

    // Sort: higher visibility score first, then proximity
    withMeta.sort((a, b) => {
      if (b.vis.score !== a.vis.score) return b.vis.score - a.vis.score;
      return a.distanceNm - b.distanceNm;
    });

    return withMeta;
  }

  /**
   * Same per-aircraft computation as build(), but with NO relevance or
   * suppression filtering — every tracked aircraft, nearest first. Used by
   * the ground-truth logging panel (src/dev/logPanel.js), which needs to
   * log "not visible" observations against aircraft the relevance filter
   * already excluded, not just the ones NAV currently shows.
   */
  function buildAll(aircraftList, userState, staleThresholdSeconds) {
    const withMeta = _computeAll(aircraftList, userState, staleThresholdSeconds);
    withMeta.sort((a, b) => a.distanceNm - b.distanceNm);
    return withMeta;
  }

  /**
   * Nudges apart indicators whose projected screen positions land too close
   * together. Now that Geo.projectToPolarPosition() scatters aircraft across
   * the whole plot by bearing AND distance rather than confining them to a
   * shared edge, two indicators can end up close in ANY direction — not just
   * along one shared axis — so this resolves plain 2D proximity: any pair
   * closer than minGapPx is pushed apart along the line between their
   * centres, split evenly between the two. A few passes handle chains (A
   * pushed into B's space gets resolved on the next pass). Mutates and
   * returns the same array; only meant to run on the already-capped/
   * paginated subset actually being rendered, not the full relevant list.
   *
   * @param {Array} items      Items with x/y, as produced by build()/buildAll().
   * @param {number} minGapPx  Minimum centre-to-centre spacing.
   */
  function declutter(items, minGapPx) {
    const MAX_PASSES = 4;

    for (let pass = 0; pass < MAX_PASSES; pass++) {
      let movedAny = false;

      for (let i = 0; i < items.length; i++) {
        for (let j = i + 1; j < items.length; j++) {
          const a = items[i], b = items[j];
          const dx = b.x - a.x, dy = b.y - a.y;
          const dist = Math.hypot(dx, dy);

          if (dist === 0) {
            // Exactly coincident — no direction to push along, so pick one.
            a.x -= minGapPx / 2;
            b.x += minGapPx / 2;
            movedAny = true;
          } else if (dist < minGapPx) {
            const push = (minGapPx - dist) / 2;
            const ux = dx / dist, uy = dy / dist;
            a.x -= ux * push; a.y -= uy * push;
            b.x += ux * push; b.y += uy * push;
            movedAny = true;
          }
        }
      }

      if (!movedAny) break;
    }

    items.forEach(item => {
      item.x = Math.round(item.x);
      item.y = Math.round(item.y);
    });

    return items;
  }

  return { build, buildAll, capForViewportWidth, declutter, POLAR_MAX_RANGE_NM, RING_BANDS_NM, FOV_HALF_ANGLE_DEG };
})();

if (typeof module !== "undefined") module.exports = Indicators;

File: android/app/src/main/java/org/vectair/vcas/car/logic/Indicators.kt
package org.vectair.vcas.car.logic

/**
 * Builds sorted, filtered indicator data for the Driving View. Ported
 * from the PWA's src/logic/indicators.js, the fifth file in CLAUDE.md's
 * "Android Auto — native rewrite scoping" pure-logic list (after
 * Geo.kt, Visibility.kt, Relevance.kt, AircraftExtrapolation.kt) — the
 * orchestration layer that ties all four of the prior ports together.
 *
 * `Aircraft` here is `AircraftExtrapolation.Aircraft` — the same
 * full-fidelity normalised domain object that port already introduced
 * (mirroring src/data/normaliseAircraft.js's real shape), reused rather
 * than yet another narrow per-file type, since this is the first
 * consumer that genuinely needs the whole thing at once: hex (for
 * suppression filtering), lat/lon/altitudeFt/type/category/
 * lastSeenSeconds (for `Visibility.estimate()`), and
 * lat/lon/altitudeFt/trackDeg/groundSpeedKt (for `Relevance.evaluate()`).
 * `computeAll()` adapts it into each dependency's own narrower input
 * type (`Visibility.AircraftInput`, `Relevance.AircraftState`) at the
 * call site, the same way the JS original hands the same duck-typed
 * object to both without either module needing to know about the
 * other's exact field list.
 *
 * Every formula, threshold, and doc-comment reasoning is preserved from
 * the original — see indicators.js itself for the fuller per-constant
 * design rationale (viewport-tiered caps, the banded polar scale, the
 * FOV restriction) — not re-duplicated here to avoid the two copies
 * drifting apart in wording even though the numbers/logic must stay in
 * sync by hand across both platforms.
 */
object Indicators {

    // Viewport-tiered NAV display cap — small phone screens stay
    // glanceable with fewer indicators; larger tablet/infotainment-style
    // displays have room for more before it becomes clutter.
    private const val NAV_CAP_SMALL_MAX_WIDTH = 500.0  // px, exclusive
    private const val NAV_CAP_MEDIUM_MAX_WIDTH = 900.0 // px, inclusive
    private const val NAV_CAP_SMALL = 5
    private const val NAV_CAP_MEDIUM = 7
    private const val NAV_CAP_LARGE = 10

    /** Max NAV indicators to show at a given viewport width (px, real or emulated). */
    fun capForViewportWidth(width: Double): Int {
        if (width < NAV_CAP_SMALL_MAX_WIDTH) return NAV_CAP_SMALL
        if (width <= NAV_CAP_MEDIUM_MAX_WIDTH) return NAV_CAP_MEDIUM
        return NAV_CAP_LARGE
    }

    /**
     * True polar plot range — the distance (nm) that maps to the full
     * available radius from the anchor. Reuses Relevance's own dead-ahead
     * teardrop range extension cap rather than a separate made-up number
     * — see indicators.js's own doc comment for the full "why" (an
     * aircraft at the plot's outer edge should mean something about
     * relevance, not just be an arbitrary cutoff).
     */
    val POLAR_MAX_RANGE_NM: Double = Relevance.DEFAULTS.rangeExtensionCapNm

    /**
     * Ring band boundaries (nm) for the plot's non-linear distance scale
     * — see Geo.bandedRadiusFraction(). See indicators.js's own doc
     * comment for the full rationale (equal screen-space slices per band
     * regardless of real nm width, a deliberately wide final band for
     * sparser high-altitude/contrail traffic).
     */
    val RING_BANDS_NM: List<Double> = listOf(2.0, 5.0, 10.0, 15.0, POLAR_MAX_RANGE_NM)

    /**
     * Half-angle (deg) of RAW mode's forward field of view — 75 deg
     * either side of dead-ahead, ~150 deg total, matching a real
     * TCAS/ND reference photo. Passed through as userState.fovHalfAngleDeg
     * by RAW's caller only.
     */
    const val FOV_HALF_ANGLE_DEG = 75.0

    /** Mirrors indicators.js's `userState` shape — see per-field comments below. */
    data class UserState(
        val lat: Double,
        val lon: Double,
        val heading: Double,
        val speedMph: Double? = null,
        val viewportWidth: Double = 0.0,
        val viewportHeight: Double = 0.0,
        val metar: Visibility.Metar? = null,
        // Must match what the camera actually used for this frame or the
        // plotted origin silently drifts from where the user marker/range
        // rings really are — see geo.js's projectToPolarPosition doc
        // comment. Null (not yet evaluated a camera frame) falls back to
        // Geo.kt's own projectToPolarPosition default (0.8), matching JS's
        // implicit undefined-passthrough exactly.
        val anchorY: Double? = null,
        val fovHalfAngleDeg: Double? = null,
        // RAW's plot lives inside a 1:1 square sub-region of the real
        // viewport (Geo.computeSquarePlotLayout), not the full viewport
        // itself — null (Hybrid, which never sets these) falls back to the
        // plain viewport with no offset.
        val plotWidth: Double? = null,
        val plotHeight: Double? = null,
        val plotOffsetX: Double? = null,
        val plotOffsetY: Double? = null,
        // "Real chrome height to stay clear of" for Hybrid's unrestricted,
        // full-viewport teardrop. Null falls back to Geo.kt's own
        // projectToPolarPosition default (60.0).
        val safeInset: Double? = null,
        // A small fixed in-square margin for RAW specifically — the square
        // already excludes all real chrome from its own bounds, so reusing
        // the full safeInset value within it would needlessly shrink an
        // already chrome-free region. Null (Hybrid) falls back to safeInset.
        val plotSafeInset: Double? = null,
        // The user-selectable ND-style range knob overrides which band
        // boundaries are "in play" for this render. Null falls back to the
        // full RING_BANDS_NM.
        val plotBandsNm: List<Double>? = null
    )

    data class IndicatorItem(
        val aircraft: AircraftExtrapolation.Aircraft,
        val bearing: Double,
        val distanceNm: Double,
        val relativeBearing: Double,
        val relativeTrackDeg: Double?,
        val vis: Visibility.EstimateResult,
        val relevance: Relevance.EvaluationResult,
        val x: Int?,
        val y: Int?,
        val isStale: Boolean
    )

    /**
     * Shared per-aircraft computation used by both build() and buildAll()
     * — bearing/distance/visibility/relevance/polar screen position/
     * direction-of-travel, for every aircraft that passes the hard
     * staleness cutoff. Does NOT filter by relevance or suppression;
     * callers decide what to do with that.
     */
    private fun computeAll(
        aircraftList: List<AircraftExtrapolation.Aircraft>,
        userState: UserState,
        staleThresholdSeconds: Double
    ): List<IndicatorItem> {
        val anchorY = userState.anchorY ?: 0.8 // Geo.kt's own projectToPolarPosition default
        val fovHalfAngleDeg = userState.fovHalfAngleDeg
        val plotWidth = userState.plotWidth ?: userState.viewportWidth
        val plotHeight = userState.plotHeight ?: userState.viewportHeight
        val plotOffsetX = userState.plotOffsetX ?: 0.0
        val plotOffsetY = userState.plotOffsetY ?: 0.0
        val safeInset = (userState.plotSafeInset ?: userState.safeInset) ?: 60.0 // Geo.kt's own default
        val bandsNm = userState.plotBandsNm ?: RING_BANDS_NM

        return aircraftList
            .filter { it.lastSeenSeconds < staleThresholdSeconds * 3 } // hard cut
            .map { a ->
                val bearing = Geo.calculateBearing(userState.lat, userState.lon, a.lat, a.lon)
                val distanceNm = Geo.calculateDistanceNm(userState.lat, userState.lon, a.lat, a.lon)
                val vis = Visibility.estimate(
                    userState.lat, userState.lon,
                    Visibility.AircraftInput(a.lat, a.lon, a.altitudeFt, a.type, a.category, a.lastSeenSeconds),
                    userState.metar
                )
                val relativeBearing = Geo.calculateRelativeBearing(bearing, userState.heading)
                val relevance = Relevance.evaluate(
                    Relevance.UserState(userState.lat, userState.lon, userState.heading, userState.speedMph),
                    Relevance.AircraftState(a.lat, a.lon, a.altitudeFt, a.trackDeg, a.groundSpeedKt),
                    relativeBearing,
                    Relevance.VisInput(vis.slantRangeNm, vis.elevationDeg)
                )
                // Slant range (not flat horizontal distance) — the same
                // figure Relevance itself compares against the teardrop, so
                // an aircraft's plotted radius agrees with whether it's near
                // the edge of relevance. null when fovHalfAngleDeg is set
                // and this bearing falls outside RAW's forward field of
                // view — callers must skip rendering those.
                val pos = Geo.projectToPolarPosition(
                    relativeBearing, vis.slantRangeNm, plotWidth, plotHeight, bandsNm,
                    anchorY, safeInset, fovHalfAngleDeg, plotOffsetX, plotOffsetY
                )
                // Direction-of-travel indicator — null when the aircraft
                // isn't transmitting a track (no arrow drawn for those).
                val relativeTrackDeg = a.trackDeg?.let { Geo.calculateRelativeBearing(it, userState.heading) }
                val isStale = a.lastSeenSeconds > staleThresholdSeconds

                IndicatorItem(
                    aircraft = a,
                    bearing = bearing,
                    distanceNm = distanceNm,
                    relativeBearing = relativeBearing,
                    relativeTrackDeg = relativeTrackDeg,
                    vis = vis,
                    relevance = relevance,
                    x = pos?.x,
                    y = pos?.y,
                    isStale = isStale
                )
            }
    }

    /**
     * Given the full aircraft list and user state, return every relevant
     * aircraft, sorted best-candidate-first. Aircraft that
     * Relevance.evaluate() rules out never reach the sort stage at all —
     * this is a TCAS-style relevance gate, not just a visibility ranking.
     *
     * Deliberately unpaginated — the caller decides how many to actually
     * display (via capForViewportWidth) and which page.
     *
     * @param suppressedHexes Aircraft hex codes to exclude regardless of
     *   relevance (manually dismissed via the popup's Suppress button).
     *   Applies uniformly — no relevance reason is exempt, including
     *   overhead/close cases.
     */
    fun build(
        aircraftList: List<AircraftExtrapolation.Aircraft>,
        userState: UserState,
        staleThresholdSeconds: Double,
        suppressedHexes: Set<String>? = null
    ): List<IndicatorItem> {
        return computeAll(aircraftList, userState, staleThresholdSeconds)
            .filter { suppressedHexes == null || !suppressedHexes.contains(it.aircraft.hex) }
            .filter { it.relevance.relevant }
            // Higher visibility score first, then proximity.
            .sortedWith(compareByDescending<IndicatorItem> { it.vis.score }.thenBy { it.distanceNm })
    }

    /**
     * Same per-aircraft computation as build(), but with NO relevance or
     * suppression filtering — every tracked aircraft, nearest first. Used
     * by the ground-truth logging panel, which needs to log "not visible"
     * observations against aircraft the relevance filter already
     * excluded, not just the ones NAV currently shows.
     */
    fun buildAll(
        aircraftList: List<AircraftExtrapolation.Aircraft>,
        userState: UserState,
        staleThresholdSeconds: Double
    ): List<IndicatorItem> {
        return computeAll(aircraftList, userState, staleThresholdSeconds).sortedBy { it.distanceNm }
    }
}
